src/librustc_metadata/cstore_impl.rs

```diff
@@ -486,6 +487,10 @@ impl CrateStore for cstore::CStore {
         self.get_crate_data(cnum).name
     }
 
+    fn crate_is_private_dep_untracked(&self, cnum: CrateNum) -> bool {
+        self.get_crate_data(cnum).private_dep
```

Comment:
The privacy flags should indeed be immediately tied to `CrateNum` on crate loading, rather than determined during the privacy pass in shady ways.
IIRC, I mentioned it during the review of https://github.com/rust-lang/rust/pull/57586, but it wasn't high-priority and was left unfixed in that PR, and I actually added a task into my work queue to fix this later.
Now I need one less thing to do, yay.